File: src/audio_stegano_app/lib/core/ui/content_text_direction.dart
import 'package:flutter/material.dart';

/// Resolves text direction for input/display based on content vs UI locale.
abstract final class ContentTextDirection {
  static final RegExp _rtlScript = RegExp(
    r'[\u0590-\u05FF\u0600-\u06FF\u0750-\u077F\u08A0-\u08FF\uFB50-\uFDFF\uFE70-\uFEFF]',
  );

  /// Latin letters (incl. extended), digits, whitespace, common Western punctuation.
  static final RegExp _latinOnly = RegExp(
    r'''^[\s0-9A-Za-z\u00C0-\u024F\u1E00-\u1EFF.,:;!?@#$%^&*()_+\-=\[\]{}|\\/<>\"'`~]*$''',
  );

  static bool containsRtlScript(String text) => _rtlScript.hasMatch(text);

  static bool isLatinOnly(String text) {
    if (text.trim().isEmpty) return false;
    if (containsRtlScript(text)) return false;
    return _latinOnly.hasMatch(text);
  }

  static TextDirection resolve(
    String text, {
    required TextDirection localeDirection,
    bool forceLatinLtr = false,
  }) {
    if (forceLatinLtr) return TextDirection.ltr;
    if (text.trim().isEmpty) return localeDirection;
    if (containsRtlScript(text)) return TextDirection.rtl;
    if (isLatinOnly(text)) return TextDirection.ltr;
    return localeDirection;
  }
}

File: src/audio_stegano_desktop/src/AudioStegano.Core/Ui/ContentTextDirection.cs
namespace AudioStegano.Core.Ui;

/// Resolves text direction for input/display: Latin-only content stays LTR in RTL UI locales.
public static class ContentTextDirection
{
    private const string RtlScriptPattern =
        @"[\u0590-\u05FF\u0600-\u06FF\u0750-\u077F\u08A0-\u08FF\uFB50-\uFDFF\uFE70-\uFEFF]";

    private const string LatinOnlyPattern =
        @"^[\s0-9A-Za-z\u00C0-\u024F\u1E00-\u1EFF.,:;!?@#$%^&*()_+\-=\[\]{}|\\/<>""'`~]*$";

    private static readonly System.Text.RegularExpressions.Regex RtlScript =
        new(RtlScriptPattern, System.Text.RegularExpressions.RegexOptions.Compiled);

    private static readonly System.Text.RegularExpressions.Regex LatinOnly =
        new(LatinOnlyPattern, System.Text.RegularExpressions.RegexOptions.Compiled);

    public static bool ContainsRtlScript(string? text) =>
        !string.IsNullOrEmpty(text) && RtlScript.IsMatch(text);

    public static bool IsLatinOnly(string? text)
    {
        if (string.IsNullOrWhiteSpace(text)) return false;
        if (ContainsRtlScript(text)) return false;
        return LatinOnly.IsMatch(text);
    }

    /// <summary>
    /// True when content should render left-to-right (Latin-only or forced).
    /// </summary>
    public static bool UseLeftToRight(string? text, bool forceLatinLtr = false, bool uiIsRtl = false)
    {
        if (forceLatinLtr) return true;
        if (string.IsNullOrWhiteSpace(text)) return !uiIsRtl;
        if (ContainsRtlScript(text)) return false;
        if (IsLatinOnly(text)) return true;
        return !uiIsRtl;
    }
}
